tbdex/quote/quote.go

```diff
@@ -0,0 +1,252 @@
+	Payin     QuoteDetails `json:"payin"`
+	Payout    QuoteDetails `json:"payout"`
+}
+
+// QuoteDetails describes the relevant information of a currency that is being sent or received
+type QuoteDetails struct {
+	CurrencyCode       string              `json:"currencyCode"`
+	Amount             string              `json:"amount"`
+	Fee                string              `json:"fee,omitempty"`
+	PaymentInstruction *PaymentInstruction `json:"paymentInstruction,omitempty"`
+}
+
+// PaymentInstruction contains instructions with plain text and/or a link
+type PaymentInstruction struct {
+	Link        string `json:"link,omitempty"`
+	Instruction string `json:"instruction,omitempty"`
+}
+
+// Digest computes a hash of the quote
+func (q Quote) Digest() ([]byte, error) {
+	payload := map[string]any{"metadata": q.MessageMetadata, "data": q.Data}
+
+	hashed, err := tbdex.DigestJSON(payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to digest quote: %w", err)
+	}
+
+	return hashed, nil
+}
+
+// Sign cryptographically signs the quote using DID's private key
+func (q *Quote) Sign(bearerDID did.BearerDID) error {
+	q.MessageMetadata.From = bearerDID.URI
+
+	signature, err := tbdex.Sign(q, bearerDID)
+	if err != nil {
+		return fmt.Errorf("failed to sign rfq: %w", err)
+	}
+
+	q.Signature = signature
+
+	return nil
+}
+
+// Verify verifies the signature of the quote.
+func (q *Quote) Verify() error {
+	_, err := tbdex.VerifySignature(q, q.Signature)
+	if err != nil {
+		return fmt.Errorf("failed to verify quote signature: %w", err)
+	}
+
+	// TODO add check when decoded.SignerDID is implemented
+	// if decoded.SignerDID != r.MessageMetadata.From {
+	// 	return errors.New("signer: %w does not match message metadata from: %w", decoded.Header.SignerDID, r.MessageMetadata.From)
+	// }
+
+	return nil
+}
+
+// UnmarshalJSON validates and unmarshals the input data into an Quote.
+func (q *Quote) UnmarshalJSON(data []byte) error {
+	err := tbdex.Validate(tbdex.TypeMessage, data, tbdex.WithKind(Kind))
+	if err != nil {
+		return fmt.Errorf("invalid quote: %w", err)
+	}
+
+	ret := quote{}
+	err = json.Unmarshal(data, &ret)
+	if err != nil {
+		return fmt.Errorf("failed to JSON unmarshal quote: %w", err)
+	}
+
+	*q = Quote(ret)
+
+	return nil
+}
+
+// Parse validates, parses input data into an RFQ, and verifies the signature and private data.
+func (q *Quote) Parse(data []byte, privateDataStrict bool) error {
+	err := q.UnmarshalJSON(data)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal Quote: %w", err)
+	}
+
+	err = q.Verify()
+	if err != nil {
+		return fmt.Errorf("failed to verify Quote: %w", err)
+	}
+
+	return nil
+}
+
+// Create generates a new Quote with the specified parameters and options.
+func Create(from, to, exchangeID, expiresAt string, payin, payout QuoteDetails, opts ...CreateOption) Quote {
+	q := createOptions{
+		id:        typeid.Must(typeid.WithPrefix(Kind)).String(),
+		createdAt: time.Now(),
+		protocol:  "1.0",
```

Comment:
would this override protocol value as `1.0` even if `protocol` value was passed in?

Reply:
oh we don't have `func Protocol()` createoption... 🤔 totally fine if we just hardcoding it too.